File: Sources/Providers/CodexLocalProvider.swift
import Foundation
import SQLite3

/// Reads live account limits using the session owned and refreshed by Codex.
actor CodexLocalProvider: UsageProvider {
    nonisolated let id = "codex"
    nonisolated let displayName = "Codex"
    nonisolated let glyph = ProviderGlyph.openai

    private let session: URLSession
    nonisolated private let authURL: URL
    private let archive: UsageArchive
    private var retryNoEarlierThan: Date?

    init(session: URLSession = .shared,
         authURL: URL = CodexCredentials.authURL,
         archive: UsageArchive = UsageArchive()) {
        self.session = session
        self.authURL = authURL
        self.archive = archive
        // Recreating the provider or relaunching must not bypass the server's retry deadline.
        self.retryNoEarlierThan = archive.loadBackoffUntil(providerID: "codex")
    }

    nonisolated var signInRoute: SignInRoute { .openApp(bundleID: "com.openai.codex", name: "Codex") }

    nonisolated func account() -> ProviderAccount? { CodexCredentials.account(from: authURL) }

    func fetchSnapshot() async throws -> ProviderSnapshot {
        let now = Date()
        if let retryNoEarlierThan, retryNoEarlierThan > now {
            throw UsageProviderError.rateLimited(retryAfter: retryNoEarlierThan.timeIntervalSince(now))
        }

        // Codex can rotate its token between polls; this app never refreshes or writes it.
        let credential = try CodexCredentials.load(from: authURL)
        var request = URLRequest(
            url: URL(string: "https://chatgpt.com/backend-api/wham/usage")!,
            cachePolicy: .reloadIgnoringLocalCacheData,
            timeoutInterval: 15
        )
        request.httpMethod = "GET"
        request.setValue("Bearer \(credential.accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue(credential.accountID, forHTTPHeaderField: "ChatGPT-Account-Id")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("no-cache, no-store", forHTTPHeaderField: "Cache-Control")

        let (data, response) = try await session.data(for: request)
        let http = response as? HTTPURLResponse
        let status = http?.statusCode ?? 0
        if status == 401 || status == 403 { throw UsageProviderError.needsAuth }
        if status == 429 {
            let receivedAt = Date()
            let delay = max(60, Self.retryAfter(from: http, now: receivedAt) ?? 0)
            retryNoEarlierThan = receivedAt.addingTimeInterval(delay)
            archive.saveBackoffUntil(retryNoEarlierThan, providerID: id)
            throw UsageProviderError.rateLimited(retryAfter: delay)
        }
        guard (200..<300).contains(status) else {
            throw UsageProviderError.badResponse(status: status)
        }

        let windows = try CodexUsage.windows(from: data)
        retryNoEarlierThan = nil
        archive.saveBackoffUntil(nil, providerID: id)
        return ProviderSnapshot(
            id: id, displayName: displayName, glyph: glyph,
            fidelity: .official, status: .ok, windows: windows,
            headlineID: windows.first?.id
        )
    }

    private static func retryAfter(from response: HTTPURLResponse?, now: Date) -> TimeInterval? {
        guard let header = response?.value(forHTTPHeaderField: "Retry-After")?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        else { return nil }
        if let seconds = TimeInterval(header), seconds.isFinite { return max(0, seconds) }

        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "GMT")
        formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss zzz"
        guard let date = formatter.date(from: header) else { return nil }
        return max(0, date.timeIntervalSince(now))
    }
}

/// Shared access to Codex's local state.
enum CodexStore {
    static var stateURL: URL {
        URL(fileURLWithPath: NSHomeDirectory()).appendingPathComponent(".codex/state_5.sqlite")
    }

    /// The desktop app's own thread catalogue.
    ///
    /// Codex's *rollouts* are written by the CLI and by the VS Code extension.
    /// The desktop app — ChatGPT.app, which is what most people mean by "Codex"
    /// now — writes none of them; it keeps its threads here instead, with
    /// `source_kind = 'chatgpt'`. Watching only the rollouts meant the notch
    /// could never see the desktop app working at all.
    static var desktopStoreURL: URL {
        URL(fileURLWithPath: NSHomeDirectory())
            .appendingPathComponent(".codex/sqlite/codex-dev.db")
    }

    /// The most recently touched desktop thread: when, and what it is called.
    static func newestDesktopThread(in url: URL) -> (title: String, updatedAt: Date)? {
        guard let db = SQLiteStore.open(url) else { return nil }
        defer { sqlite3_close(db) }

        let rows = SQLiteStore.rows(
            in: db,
            sql: """
            SELECT source_updated_at, display_title, thread_id
            FROM local_thread_catalog ORDER BY source_updated_at DESC LIMIT 1
            """,
            columns: 3
        )
        guard let row = rows.first, let seconds = Double(row[0]) else { return nil }
        // Seconds since the epoch, with a fractional part — not the
        // milliseconds the `threads` table next door uses.
        let title = row[1].isEmpty ? "Codex" : row[1]
        return (title, Date(timeIntervalSince1970: seconds))
    }

    /// The rollout of the most recently touched thread.
    static func newestRollout(in store: URL) -> URL? {
        guard let db = SQLiteStore.open(store) else { return nil }
        defer { sqlite3_close(db) }

        let paths = SQLiteStore.rows(
            in: db,
            sql: "SELECT rollout_path FROM threads WHERE archived = 0 ORDER BY updated_at_ms DESC LIMIT 8"
        )
        return paths
            .map { URL(fileURLWithPath: ($0 as NSString).expandingTildeInPath) }
            .first { FileManager.default.fileExists(atPath: $0.path) }
    }
}

File: windows/QuotaArc/Providers/CodexLocalProvider.cs
using System.Net.Http.Headers;
using QuotaArc.Model;

namespace QuotaArc.Providers;

internal sealed class CodexLocalProvider : IUsageProvider
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(15) };
    private readonly UsageArchive _archive;
    private DateTime? _retryNoEarlierThan;

    public CodexLocalProvider(UsageArchive? archive = null)
    {
        _archive = archive ?? new UsageArchive();
        _retryNoEarlierThan = _archive.LoadBackoffUntil(Id);
    }

    public string Id => "codex";
    public string DisplayName => "Codex";
    public ProviderGlyph Glyph => ProviderGlyph.Openai;
    public SignInRoute SignInRoute => new SignInRoute.OpenApp("Codex");
    public ProviderAccount? Account() => CodexCredentials.Account();

    public async Task<ProviderSnapshot> FetchSnapshotAsync()
    {
        var now = DateTime.Now;
        if (_retryNoEarlierThan is { } until && until > now)
            throw UsageProviderException.RateLimited((until - now).TotalSeconds);

        var credential = CodexCredentials.Load();
        using var req = new HttpRequestMessage(HttpMethod.Get, "https://chatgpt.com/backend-api/wham/usage");
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", credential.AccessToken);
        req.Headers.TryAddWithoutValidation("ChatGPT-Account-Id", credential.AccountId);
        req.Headers.TryAddWithoutValidation("Accept", "application/json");
        req.Headers.CacheControl = new CacheControlHeaderValue { NoCache = true, NoStore = true };
        using var res = await Http.SendAsync(req);
        var status = (int)res.StatusCode;
        if (status is 401 or 403) throw UsageProviderException.NeedsAuth();
        if (status == 429)
        {
            var delay = Math.Max(60, RetryAfter(res) ?? 0);
            _retryNoEarlierThan = DateTime.Now.AddSeconds(delay);
            _archive.SaveBackoffUntil(_retryNoEarlierThan, Id);
            throw UsageProviderException.RateLimited(delay);
        }
        if (status is < 200 or >= 300) throw UsageProviderException.BadResponse(status);

        var data = await res.Content.ReadAsByteArrayAsync();
        var windows = CodexUsage.WindowsFrom(data);
        _retryNoEarlierThan = null;
        _archive.SaveBackoffUntil(null, Id);
        return new ProviderSnapshot(
            Id, DisplayName, Glyph, Fidelity.Official, new ProviderStatus.Ok(),
            windows, windows.FirstOrDefault()?.Id);
    }

    private static double? RetryAfter(HttpResponseMessage res)
    {
        if (res.Headers.RetryAfter?.Delta is { } d) return Math.Max(0, d.TotalSeconds);
        return null;
    }
}
